*************************************************************************************************
From: Karunakaran Amirthalingam
Date: Thu, 8 Jan 2026 10:49:57 +0000
Subject: [PATCH] 0004-Add-support-for-project-dir.patch
Reason for Change: This patch is to add support for PROJECTDIR support to ThunderTools
Upstream-Status: Comcast changes
Signed-off-by: Karunakaran Amirthalingam <karunakaran_amirthalingam@cable.comcast.com>
*****************************************************************************************************
diff --git a/JsonGenerator/source/class_emitter.py b/JsonGenerator/source/class_emitter.py
index 187a0a3..b7bdff6 100644
--- a/JsonGenerator/source/class_emitter.py
+++ b/JsonGenerator/source/class_emitter.py
@@ -251,8 +251,6 @@ def EmitEnumRegs(log, root, emit, header_file, if_file):
     emit.Line("#include <core/Enumerate.h>")
     emit.Line()

-    emit.Line("#include \"definitions.h\"")
-
     if not config.NO_INCLUDES:
         if if_file.endswith(".h"):
             emit.Line("#include <%s%s>" % (config.CPP_INTERFACE_PATH, if_file))
diff --git a/ProxyStubGenerator/StubGenerator.py b/ProxyStubGenerator/StubGenerator.py
index e183af6..04a2498 100755
--- a/ProxyStubGenerator/StubGenerator.py
+++ b/ProxyStubGenerator/StubGenerator.py
@@ -205,7 +205,7 @@ def Flatten(identifier, scope):
 
 
 # Generate interface information in lua
-def GenerateLuaData(emit, interfaces_list, enums_list, source_file=None, tree=None, ns=None):
+def GenerateLuaData(emit, interfaces_list, enums_list, project_dir, source_file=None, tree=None, ns=None):
 
     if not source_file:
         assert tree==None
@@ -567,7 +567,7 @@ def Parse(source_file, framework_namespace, includePaths = [], defaults = "", ex
 
     return tree
 
-def GenerateStubs2(output_file, source_file, tree, ns, scan_only=False):
+def GenerateStubs2(output_file, source_file, project_dir, tree, ns, scan_only=False):
     log.Info("Scanning '%s' (in %s)..." % (source_file, ns))
 
     if not FORCE and (os.path.exists(output_file) and (os.path.getmtime(source_file) < os.path.getmtime(output_file))):
@@ -2532,6 +2532,8 @@ def GenerateStubs2(output_file, source_file, tree, ns, scan_only=False):
 
         if os.path.isfile(os.path.join(os.path.dirname(source_file), "Module.h")):
             emit.Line('#include "Module.h"')
+        elif os.path.isfile(os.path.join(project_dir, "Module.h")):
+            emit.Line('#include "Module.h"')
 
         if os.path.isfile(os.path.join(os.path.dirname(source_file), interface_header_name)):
             emit.Line('#include "%s"' % interface_header_name)
@@ -2720,6 +2722,8 @@ if __name__ == "__main__":
                            help="include an additional C++ header file, can be used multiple times (default: include 'Ids.h')")
     argparser.add_argument('-I', dest="includePaths", metavar="INCLUDE_DIR", action='append', default=[], type=str,
                            help='add an include search path, can be used multiple times')
+    argparser.add_argument("--projectdir",dest="project_dir",metavar="DIR",type=str,default="",
+                           help="specify the project directory")
 
     args = argparser.parse_args(sys.argv[1:])
     SHOW_WARNINGS = not args.no_warnings
@@ -2870,7 +2883,11 @@ if __name__ == "__main__":
 
             for source_file in interface_files:
                 try:
-                    _extra_includes = [ os.path.join("@" + os.path.dirname(source_file), IDS_DEFINITIONS_FILE) ]
+                    if os.path.isfile(os.path.join(args.project_dir, IDS_DEFINITIONS_FILE)):
+                        _extra_includes = [ os.path.join("@" + args.project_dir, IDS_DEFINITIONS_FILE) ]
+                    else:
+                        _extra_includes = [ os.path.join("@" + os.path.dirname(source_file), IDS_DEFINITIONS_FILE) ]
+
                     _extra_includes.extend(args.extra_includes)
 
                     tree = Parse(source_file, FRAMEWORK_NAMESPACE, args.includePaths,
@@ -2891,7 +2908,7 @@ if __name__ == "__main__":
                         some_omitted = False
 
                         for ns in INTERFACE_NAMESPACES:
-                            output, some_omitted = GenerateStubs2(output_file, source_file, tree, ns, scan_only)
+                            output, some_omitted = GenerateStubs2(output_file, source_file, args.project_dir, tree, ns, scan_only)
 
                             new_faces += output
 
@@ -2914,7 +2931,7 @@ if __name__ == "__main__":
                         log.Info("(lua generator) Scanning %s..." % os.path.basename(source_file))
 
                         for ns in INTERFACE_NAMESPACES:
-                            GenerateLuaData(Emitter(lua_file, INDENT_SIZE), lua_interfaces, lua_enums, source_file, tree, ns)
+                            GenerateLuaData(Emitter(lua_file, INDENT_SIZE), lua_interfaces, lua_enums, args.project_dir, source_file, tree, ns)
 
                 except NotModifiedException as err:
                     log.Info("skipped file %s, up-to-date" % os.path.basename(output_file))
@@ -2965,7 +2982,7 @@ if __name__ == "__main__":
             if args.lua_code:
                 # Epilogue
                 for ns in INTERFACE_NAMESPACES:
-                    GenerateLuaData(Emitter(lua_file, INDENT_SIZE), lua_interfaces, lua_enums)
+                    GenerateLuaData(Emitter(lua_file, INDENT_SIZE), lua_interfaces, lua_enums, args.project_dir)
                     log.Info("Created %s (%s interfaces, %s enums)" % (lua_file.name, len(lua_interfaces), len(lua_enums)))
 
         else:
diff --git a/cmake/FindProxyStubGenerator.cmake.in b/cmake/FindProxyStubGenerator.cmake.in
index 1615b79..89d5bc6 100644
--- a/cmake/FindProxyStubGenerator.cmake.in
+++ b/cmake/FindProxyStubGenerator.cmake.in
@@ -31,7 +31,7 @@ function(ProxyStubGenerator)
     endif()
 
     set(optionsArgs SECURE COHERENT TRACES VERBOSE NO_WARNINGS KEEP FORCE_GENERATE COLLATED_ITERATORS)
-    set(oneValueArgs OUTDIR FRAMEWORK_NAMESPACE)
+    set(oneValueArgs OUTDIR FRAMEWORK_NAMESPACE PROJECT_DIR)
     set(multiValueArgs INPUT INCLUDE INCLUDE_PATH NAMESPACE)
 
     cmake_parse_arguments(Argument "${optionsArgs}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN} )
@@ -85,6 +85,10 @@ function(ProxyStubGenerator)
         list(APPEND _execute_command  "--outdir" "${Argument_OUTDIR}")
     endif()
 
+    if(Argument_PROJECT_DIR)
+       list(APPEND _execute_command  "--projectdir" "${Argument_PROJECT_DIR}")
+    endif()
+
     foreach(_namespace ${Argument_NAMESPACE})
         list(APPEND _execute_command  "--namespace" "${_namespace}")
     endforeach(_namespace)
